File: backend/atlas_opt_core/diagnostics.py
"""Structured diagnostic helpers for Atlas core responses."""

from __future__ import annotations

from typing import Any


def structured_diagnostic(
    message: str,
    severity: str = "warning",
    source: str = "compiler",
    target_id: str | None = None,
    target_kind: str = "project",
    suggested_fix: str | None = None,
    related_ids: list[str] | None = None,
) -> dict[str, Any]:
    """Return a structured diagnostic while keeping legacy fields."""

    normalized_severity = severity if severity in {"error", "warning", "info"} else "warning"
    return {
        "id": f"diagnostic-{abs(hash((message, target_id, source))) % 10_000_000}",
        "severity": normalized_severity,
        "level": normalized_severity,
        "source": source,
        "targetId": target_id,
        "sourceId": target_id,
        "targetKind": target_kind,
        "message": message,
        "suggestedFix": suggested_fix,
        "relatedIds": related_ids or [],
    }


def normalize_response_diagnostics(response: dict[str, Any], source: str = "compiler") -> dict[str, Any]:
    """Normalize response diagnostics in-place-compatible form."""

    diagnostics = response.get("diagnostics", [])
    if not isinstance(diagnostics, list):
        response["diagnostics"] = []
        return response
    response["diagnostics"] = [normalize_diagnostic(item, index, source) for index, item in enumerate(diagnostics)]
    return response


def normalize_diagnostic(item: Any, index: int, source: str) -> dict[str, Any]:
    """Normalize legacy strings/dicts to Atlas structured diagnostics."""

    if isinstance(item, str):
        return structured_diagnostic(item, "error" if "error" in item.lower() else "warning", source)
    if isinstance(item, dict):
        severity = item.get("severity") or item.get("level") or "warning"
        message = str(item.get("message") or f"Diagnostic {index}")
        return {
            **structured_diagnostic(
                message,
                severity,
                str(item.get("source") or source),
                item.get("targetId") or item.get("sourceId"),
                str(item.get("targetKind") or "project"),
                item.get("suggestedFix"),
                item.get("relatedIds") if isinstance(item.get("relatedIds"), list) else [],
            ),
            **item,
        }
    return structured_diagnostic(f"Diagnostic {index}", "warning", source)

File: apps/web/src/atlas/core/diagnostics.ts
export type AtlasDiagnosticSeverity = "error" | "warning" | "info";
export type AtlasDiagnosticSource = "frontend" | "compiler" | "cvxpy" | "solver" | "migration";
export type AtlasDiagnosticTargetKind = "modelObject" | "workspaceNode" | "connection" | "project";

export type AtlasStructuredDiagnostic = {
  id: string;
  severity: AtlasDiagnosticSeverity;
  source: AtlasDiagnosticSource;
  targetId?: string | null;
  targetKind: AtlasDiagnosticTargetKind;
  message: string;
  suggestedFix?: string;
  relatedIds: string[];
};

export function normalizeDiagnostic(
  diagnostic: string | { level?: string; severity?: string; message?: string; sourceId?: string | null; source?: string; targetId?: string | null; targetKind?: string; suggestedFix?: string; relatedIds?: unknown[] },
  index = 0,
  fallbackSource: AtlasDiagnosticSource = "frontend"
): AtlasStructuredDiagnostic {
  if (typeof diagnostic === "string") {
    return {
      id: `diagnostic-${index}`,
      severity: diagnostic.toLowerCase().includes("error") ? "error" : "warning",
      source: fallbackSource,
      targetId: null,
      targetKind: "project",
      message: diagnostic,
      relatedIds: []
    };
  }
  const severity = diagnostic.severity ?? diagnostic.level;
  return {
    id: `diagnostic-${index}`,
    severity: severity === "error" || severity === "warning" || severity === "info" ? severity : "warning",
    source: isDiagnosticSource(diagnostic.source) ? diagnostic.source : fallbackSource,
    targetId: diagnostic.targetId ?? diagnostic.sourceId ?? null,
    targetKind: isTargetKind(diagnostic.targetKind) ? diagnostic.targetKind : "project",
    message: diagnostic.message ?? "Diagnostic returned without message.",
    suggestedFix: diagnostic.suggestedFix,
    relatedIds: Array.isArray(diagnostic.relatedIds)
      ? diagnostic.relatedIds.filter((id): id is string => typeof id === "string")
      : []
  };
}

function isDiagnosticSource(value: unknown): value is AtlasDiagnosticSource {
  return value === "frontend" || value === "compiler" || value === "cvxpy" || value === "solver" || value === "migration";
}

function isTargetKind(value: unknown): value is AtlasDiagnosticTargetKind {
  return value === "modelObject" || value === "workspaceNode" || value === "connection" || value === "project";
}
